### PR TITLE
Refactoring map page to make a better use of state control and context

### PR DESCRIPTION
**Pull request by RodrigoEC**
## What this PR does?

I added a context that controls all the data that flows between components in the map page. With the context we don't need to share state through the components I just added a context that contains all this mutual data and the components that are under it's umbrella can access the state's data.

## How to test it?

This PR changes the code but hopefully doesn't add any bugs, so to test it I think that you can make the normal use of the map's page and section and see if it's working fine.

## Good material on context and useMemo

- [Context](https://react.dev/learn/passing-data-deeply-with-context#step-1-create-the-context);
- [useMemo](https://react.dev/reference/react/useMemo)
- [useMemo](https://www.freecodecamp.org/news/how-to-work-with-usememo-in-react/)

### CODE REVIEW

File: src/components/MapsMenu/MapsMenu.styles.tsx [hunk 1]
```diff
@@ -25,22 +25,6 @@ export const ItemWrapper = styled.div`
...
-  transition: 300ms;
-  cursor: pointer;
-
-  &:hover {
-    opacity: 0.6;
-    transform: scale(0.97);
-  }
-`;
-
 export const NoDataContainer = styled.div`
```
Comment (cilasmarques) [inline src/components/MapsMenu/MapsMenu.styles.tsx]: Remove unused NoDataContainer and NoDataElement

File: src/app/map/page.tsx
Comment (LazaroQueiroz) [inline src/app/map/page.tsx:1]: very nice refacts... the variables names were a bit confusing before!!

File: src/app/page.tsx
Comment (LazaroQueiroz) [inline src/app/page.tsx:1]: very nice changes!!! the suspense tag is very important to the site responsivity!!!!

File: src/components/CollabTable/CollabTable.tsx
Comment (LazaroQueiroz) [inline src/components/CollabTable/CollabTable.tsx:1]: thanks for cleaning the unnecessary consoles!

File: src/components/DateInput/DateInput.tsx
Comment (LazaroQueiroz) [inline src/components/DateInput/DateInput.tsx:1]: very nice use of useContext and useMemo!!

File: src/components/MapDescription/MapDescription.tsx
Comment (LazaroQueiroz) [inline src/components/MapDescription/MapDescription.tsx:1]: again, very nice use of useContext()!!

File: src/components/MapLegend/MapLegend.styles.tsx
Comment (LazaroQueiroz) [inline src/components/MapLegend/MapLegend.styles.tsx:1]: liked the addition of .details!! very handy!!

Comment (LazaroQueiroz) [reply src/components/MapLegend/MapLegend.styles.tsx:1]: summary too!!

File: src/components/MapLegend/MapLegend.tsx
Comment (LazaroQueiroz) [inline src/components/MapLegend/MapLegend.tsx:1]: again, useContext saving the day!!

File: src/components/MapTiff/MapTiff.tsx
Comment (LazaroQueiroz) [inline src/components/MapTiff/MapTiff.tsx:1]: very good!!

File: src/components/MapTiff/Section/MapSection.tsx
Comment (LazaroQueiroz) [inline src/components/MapTiff/Section/MapSection.tsx:1]: way cleaner!

File: src/components/TagButtonMaps/TagButtonMaps.tsx
Comment (LazaroQueiroz) [inline src/components/TagButtonMaps/TagButtonMaps.tsx:1]: useContext saving the day, again!

File: src/contexts/MapContext.tsx
Comment (LazaroQueiroz) [inline src/contexts/MapContext.tsx:1]: this code is very handy and important to the project!! very nice!!

**Review by LazaroQueiroz — APPROVED** [review]
LGTM!

File: src/components/MapsSection/MapsSection.tsx [hunk 2]
```diff
@@ -104,40 +68,25 @@ const MapsSection = ({
           <ButtonsWrapper>
             <ExpandBox
               href={`/map?name=${currentVisu.id}`}
```
Comment (cilasmarques) [inline src/components/MapsSection/MapsSection.tsx]: Fix the redirection by name
```suggestion
              href={`/map?id=${currentVisu.id}`}
```

File: src/components/MapTiff/MapTiff.tsx [hunk 3]
```diff
@@ -251,37 +243,36 @@ const MapTiff = ({
...
-        };
-      }
-    },
-    // eslint-disable-next-line react-hooks/exhaustive-deps
-    [name, year],
-  );
+      return () => {
+        cleanOcaLayers(map);
+      };
+    }
```
Comment (cilasmarques) [inline src/components/MapTiff/MapTiff.tsx]: You can use the currentVisu for the year too and avoid an unnecessary request
```
  useEffect(() => {
    if (map && name && year) {
      if (!isReduced) addPopupEffect(name, year);
      loadMapLayer(name, year);

      return () => {
        cleanOcaLayers(map);
      };
    }
  }, [name, year, map]);
```

File: src/contexts/MapContext.tsx [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+    const { name, description, imageData } =
+      tiffs.find((tiff) => tiff.fields.id === id)?.fields || defaultEEInfo;
+
+    if (!("general" in imageData) && !year) {
+      Object.entries(imageData).forEach(([currentYear, yearInfo]) => {
+        if (yearInfo.default) year = currentYear;
+      });
+
+      setCurrentVisu({ id, year });
+    }
```
Comment (cilasmarques) [inline src/contexts/MapContext.tsx:84]: You can add the year verification were
```suggestion
    } else if (!year) {
      setCurrentVisu({ id, year: "general" });
    }

    params.set("id", id);
    if (year !== "general") {
      params.set("year", year);
    } else {
      params.delete("year");
    }
```

**Review by cilasmarques — CHANGES_REQUESTED** [review]
VERY NICE!!!!!!!!!!!!!!!!!!

**Review by LeilaFarias — APPROVED** [review]
what an incredible job, you deserve the rest of the month off!

image

**Review by cilasmarques — APPROVED** [review]

**Pull request merged**